plugins/inputs/ctrlx_datalayer_sse/ctrlx_datalayer_sse.go

```diff
@@ -0,0 +1,498 @@
+package ctrlx_datalayer_sse
+
+import (
+	"bytes"
+	"crypto/tls"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/google/uuid"
+	"github.com/influxdata/telegraf"
+	"github.com/influxdata/telegraf/config"
+	"github.com/influxdata/telegraf/internal/choice"
+	"github.com/influxdata/telegraf/plugins/inputs"
+)
+
+// This plugin is based on the official ctrlX CORE API. Documentation can be found in OpenAPI format at:
+// https://boschrexroth.github.io/rest-api-description/ctrlx-automation/ctrlx-core/
+// Used APIs are:
+// * ctrlX CORE - Authorization and Authentication API
+// * ctrlX CORE - Data Layer API
+//
+// All communication between the device and this input plugin is based
+// on https REST and HTML5 Server Sent Events (sse).
+
+const (
+	defaultKeepaliveInterval          = config.Duration(60 * time.Second)
+	defaultErrorInterval              = config.Duration(10 * time.Second)
+	defaultReconnectInterval          = config.Duration(10 * time.Second)
+	defaultCreateSubscriptionInterval = config.Duration(2 * time.Second)
+	defaultPublishInterval            = config.Duration(1 * time.Second)
+	defaultSamplingInterval           = config.Duration(1 * time.Second)
+	defaultTimeout                    = config.Duration(5 * time.Second)
+	defaultQueueSize                  = 10
+	defaultQueueBehaviour             = "DiscardOldest"
+	defaultValueChange                = "StatusValue"
+	defaultMeasurementName            = "ctrlx"
+)
+
+var queueBehaviours = []string{"", "DiscardOldest", "DiscardNewest"}
+var valueChanges = []string{"", "Status", "StatusValue", "StatusValueTimestamp"}
+
+type ctrlXDataLayerSSE struct {
+	acc                telegraf.Accumulator
+	connection         *http.Client
+	token              token
+	autoReconnect      bool
+	Hostname           string          `toml:"hostname"`
+	Username           string          `toml:"username"`
+	Password           string          `toml:"password"`
+	InsecureSkipVerify bool            `toml:"insecure_skip_verify"`
+	Timeout            config.Duration `toml:"timeout"`
+	Subscription       []subscription
+	Log                telegraf.Logger `toml:"-"`
+	wg                 sync.WaitGroup
+	mux                sync.Mutex
+}
+
+type subscription struct {
+	index             int
+	Nodes             []string        `toml:"nodes"`
+	Measurement       string          `toml:"measurement"`
+	Enable            bool            `toml:"enable"`
+	PublishInterval   config.Duration `toml:"publish_interval"`
+	KeepaliveInterval config.Duration `toml:"keep_alive_interval"`
+	ErrorInterval     config.Duration `toml:"error_interval"`
+	SamplingInterval  config.Duration `toml:"sampling_interval"`
```

Comment:
See above. Maybe grouping subscriptions with identical settings and using multiple plugin instances would also suffice? If we keep this it will make-up for a very extensive config file...

Reply:
All of these settings are optional and we have reasonable default values, we don't think the customer need to set all of them, but it is necessary to have all of them in case they are needed.

Reply:
Fair enough. If you really need this level of granularity...